### PR TITLE
🦖 - Add ban/unban modules  + translations.

### PR DESCRIPTION
**Pull request by ekaliclua**

### CODE REVIEW

File: src/localization/locales/lang_hu.lua [hunk 1]
```diff
@@ -29,4 +29,7 @@ _Literals = {
     CONNECTION_CHARACTER_SELECTION_BUTTON = "Játék vele %s",
 
     CONNECTING_CREATING_CHARACTER = "Karakter létrehozása...",
+
+    KICK_DEFAULT_MESSAGE = "Ki lettél kikerülve a szerverről.",
+    BAN_DEFAULT_MESSAGE = "A szerver tiltották a kapcsolatod.",
```
Comment (CsokiHUN) [inline src/localization/locales/lang_hu.lua:34]:
```suggestion
    KICK_DEFAULT_MESSAGE = "Ki lettél dobva a szerverről.",
    BAN_DEFAULT_MESSAGE = "Ki lettél tiltva a szerverről.",
```

File: src/api/strings/shared/random.lua [hunk 2]
```diff
@@ -0,0 +1,14 @@
+---randomString
+---@param length number
+---@return string
+---@public
+function API_Strings:randomString(length)
+    local charset <const> = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+    length = length or 10
+    local randomString = ""
+    for i = 1, length do
```
Comment (CsokiHUN) [inline src/api/strings/shared/random.lua:9]:
```suggestion
    for i = 1, (length or 10) do
```

**Review by CsokiHUN — CHANGES_REQUESTED** [review]

File: src/localization/locales/lang_hu.lua (hunk 1 above)
Comment (ekaliclua) [reply src/localization/locales/lang_hu.lua:34]: copilot lol, thanks
that's correct ?     UNBAN_DEFAULT_MESSAGE = "A szerver engedélyezte a kapcsolatod.",

File: src/api/strings/shared/random.lua (hunk 2 above)
Comment (ekaliclua) [reply src/api/strings/shared/random.lua:9]: next pr

File: src/localization/locales/lang_hu.lua (hunk 1 above)
Comment (CsokiHUN) [reply src/localization/locales/lang_hu.lua:34]: the correct is 
```
UNBAN_DEFAULT_MESSAGE = "A kitiltásod feloldva.",
```

Comment (ekaliclua) [reply src/localization/locales/lang_hu.lua:34]: ok thx

File: src/localization/locales/lang_en.lua [hunk 3]
```diff
@@ -35,5 +35,7 @@ _Literals = {
     CONNECTING_CREATING_CHARACTER = "Creating character...",
 
     KICK_DEFAULT_MESSAGE = "You have been kicked from the server.",
+    BAN_DEFAULT_MESSAGE = "You have been banned from the server.",
+    UNBAN_DEFAULT_MESSAGE = "You have been unbanned from the server.",
```
Comment (MoskalykA) [inline src/localization/locales/lang_en.lua:40]:
```suggestion
```

File: src/api/strings/shared/random.lua [hunk 4]
```diff
@@ -4,9 +4,7 @@
 ---@public
 function API_Strings:randomString(length)
     local charset <const> = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
-    length = length or 10
-    local randomString = ""
-    for i = 1, length do
+    for i = 1, (length or 10) do
```
Comment (CsokiHUN) [inline src/api/strings/shared/random.lua:7]: Sorry i removed randomString variable
```suggestion
    local randomString = ""
    for i = 1, (length or 10) do
```

**Review by CsokiHUN — CHANGES_REQUESTED** [review]

File: src/modules/players/server/events/onPlayerConnecting.lua [hunk 5]
```diff
@@ -17,6 +17,22 @@ AddEventHandler("playerConnecting", function(playerName, _, connection)
...
+            MOD_Sanctions:unbanPlayer(identifier, banP.ban_id)
+            Wait(1500)
+            connection.done()
+        else
+            local timeDate <const> = API_Maths:getDateByUnixTime(banP["time"])
+            local unbanDate <const> = ("%s/%s/%s [%sh%s]"):format(timeDate.day, timeDate.month, timeDate.year, timeDate.hour, timeDate.min)
+            connection.done(("💢 %s\n\nID: %s\nReason: %s\nUnban Date: %s"):format(_Literals.BAN_DEFAULT_MESSAGE, banP["ban_id"], banP["reason"], unbanDate))
+            return
+        end
+    end
```
Comment (MoskalykA) [inline src/modules/players/server/events/onPlayerConnecting.lua]:
```suggestion
    if (MOD_Sanctions:isPlayerBan(identifier)) then
        local ban <const> = MOD_Sanctions.List.Bans[identifier]
        local currentTime <const> = os.time()
        if (currentTime > ban.time) then
            connection.update(("💢 %s"):format(_Literals.UNBAN_DEFAULT_MESSAGE))
            MOD_Sanctions:unbanPlayer(identifier, ban.ban_id)
            Wait(1500)
            connection.done()
        else
            local timeDate <const> = API_Maths:getDateByUnixTime(ban.time)
            local unbanDate <const> = ("%s/%s/%s [%sh%s]"):format(timeDate.day, timeDate.month, timeDate.year, timeDate.hour, timeDate.min)
            connection.done(("💢 %s\n\nID: %s\nReason: %s\nUnban Date: %s"):format(_Literals.BAN_DEFAULT_MESSAGE, ban.ban_id, ban.reason, unbanDate))
            return
        end
    end
```

File: src/modules/sanctions/server/functions/banPlayerID.lua [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+---@param reason string
+---@param time table
+---@return void
+function MOD_Sanctions:banPlayerID(playerSrc, reason, time)
+    if (not playerSrc or not reason or not time) then
+        return
+    end
+
+    local license <const> = API_Player:getLicense(playerSrc)
+
```
Comment (MoskalykA) [inline src/modules/sanctions/server/functions/banPlayerID.lua:12]:
```suggestion
```

File: src/modules/sanctions/server/functions/unbanPlayer.lua [hunk 7]
```diff
@@ -0,0 +1,20 @@
+---unbanPlayer
+---@param license string
+---@param ban_id string
+---@return void
+function MOD_Sanctions:unbanPlayer(license, ban_id)
+    if (not license or not ban_id) then
+        return
```
Comment (MoskalykA) [inline src/modules/sanctions/server/functions/unbanPlayer.lua:7]:
```suggestion
        return NCS:die("The license or id of the ban is missing")
```

File: src/modules/sanctions/server/functions/unbanPlayer.lua [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    end
+
+    if (not MOD_Sanctions:isPlayerBan(license)) then
+        return
```
Comment (MoskalykA) [inline src/modules/sanctions/server/functions/unbanPlayer.lua:11]:
```suggestion
        return NCS:die("This player is not banned")
```

File: src/modules/sanctions/server/functions/isPlayerBan.lua [hunk 9]
```diff
@@ -0,0 +1,14 @@
+---isPlayerBan
+---@param license string
+---@return boolean
+function MOD_Sanctions:isPlayerBan(license)
+    if (not license) then
+        return (false)
```
Comment (MoskalykA) [inline src/modules/sanctions/server/functions/isPlayerBan.lua:6]:
```suggestion
        return NCS:die("The license is missing")
```

File: src/modules/sanctions/server/functions/banPlayerID.lua [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    if (not license) then
+        return
```
Comment (MoskalykA) [inline src/modules/sanctions/server/functions/banPlayerID.lua:14]:
```suggestion
        return NCS:die("The license is missing")
```

File: src/modules/sanctions/server/functions/banPlayerID.lua [hunk 11]
```diff
@@ -0,0 +1,25 @@
+---banPlayerID
+---@param playerSrc number
+---@param reason string
+---@param time table
+---@return void
+function MOD_Sanctions:banPlayerID(playerSrc, reason, time)
+    if (not playerSrc or not reason or not time) then
+        return
```
Comment (MoskalykA) [inline src/modules/sanctions/server/functions/banPlayerID.lua:8]:
```suggestion
        return NCS:die("The player or the reason or the time is missing")
```

File: src/modules/sanctions/server/functions/banPlayer.lua [hunk 12]
```diff
@@ -0,0 +1,34 @@
+---banPlayer
+---@param license string
+---@param reason string
+---@param time table
+---@return void
+function MOD_Sanctions:banPlayer(license, reason, time)
+    if (not license or not reason or not time) then
+        return
```
Comment (MoskalykA) [inline src/modules/sanctions/server/functions/banPlayer.lua:8]:
```suggestion
        return NCS:die("The player or the reason or the time is missing")
```

File: src/modules/sanctions/server/functions/banPlayerID.lua [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+    local license <const> = API_Player:getLicense(playerSrc)
+
+    if (not license) then
+        return
+    end
+
+    if (MOD_Sanctions:isPlayerBan(license)) then
+        NCS:coreTrace(("Player %s is already banned."):format(license))
+        return
```
Comment (MoskalykA) [inline src/modules/sanctions/server/functions/banPlayerID.lua:19]:
```suggestion
        return NCS:die("The player is already banned")
```

File: src/modules/sanctions/server/functions/banPlayer.lua [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    end
+
+    if (MOD_Sanctions:isPlayerBan(license)) then
+        NCS:coreTrace(("Player %s is already banned."):format(license))
+        return
```
Comment (MoskalykA) [inline src/modules/sanctions/server/functions/banPlayer.lua:13]:
```suggestion
        return NCS:die("The player is already banned")
```

File: src/modules/sanctions/server/functions/banPlayer.lua [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    time = API_Maths:unformatTime(time)
+    time = API_Maths:addTimeToUnix(time)
+    local ban_id <const> = ("BAN-%s-%s"):format(API_Strings:randomString(3), API_Strings:randomString(5))
+
+    local query <const> = ("INSERT INTO ncs_bans (ban_id, license, reason, time) VALUES ('%s', '%s', '%s', '%i')"):format(ban_id, license, reason, time)
+    API_Database:query(query, {}, function(result)
+        if (not result) then
+            NCS:coreTrace(("Failed to insert ban for player %s."):format(license))
+            return
+        end
```
Comment (MoskalykA) [inline src/modules/sanctions/server/functions/banPlayer.lua:25]: @CsokiHUN What do you think?

File: src/api/strings/shared/random.lua [hunk 16]
```diff
@@ -0,0 +1,12 @@
+---randomString
+---@param length number
+---@return string
+---@public
+function API_Strings:randomString(length)
+    local charset <const> = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
```
Comment (MoskalykA) [inline src/api/strings/shared/random.lua:6]:
```suggestion
---randomString
---@param length number
---@param randomTemplate string
---@return string
---@public
function API_Strings:randomString(length, randomTemplate)
    local charset <const> = (randomTemplate or "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
```

**Review by MoskalykA — CHANGES_REQUESTED** [review]

File: src/modules/sanctions/server/functions/banPlayerID.lua (hunk 6 above)
Comment (ekaliclua) [reply src/modules/sanctions/server/functions/banPlayerID.lua:12]: ?

File: src/modules/sanctions/server/functions/banPlayer.lua (hunk 15 above)
Comment (CsokiHUN) [reply src/modules/sanctions/server/functions/banPlayer.lua:24]: its good, but why not store all identifiers?

Comment (ekaliclua) [reply src/modules/sanctions/server/functions/banPlayer.lua:24]: > its good, but why not store all identifiers?

bc the license was the only one identifier can't change ( if he changed rockstar account it's change btw )

Comment (MoskalykA) [reply src/modules/sanctions/server/functions/banPlayer.lua:24]: > its good, but why not store all identifiers?

It's not that, the "return" leaves me confused

**Review by CsokiHUN — APPROVED** [review]

**Review by MoskalykA — APPROVED** [review]

**Pull request merged**